### ISSUE TITLE
Defaulted constructors: Wrong constructor is chosen among multiple available constructors

### ISSUE THREAD

**Issue by usx95** [llvm/llvm-project#57046]
```c++
#include<assert.h>
template<unsigned N>
struct Foo {
  Foo() requires (N==1) { assert(0);}
  Foo() requires (N==2) = default;
};

template <typename T, unsigned N>
struct S {
  T data;
  S() requires (N==1) { assert(0); }
  S() requires (N==2) = default;
};

int main() {
  S<Foo<1>, 2> s1;
  return 0;
}
```
**Expected**: User defined constructor for `Foo<1>` should be chosen leading to assertion failure. 
**Actual**: Clang currently chooses the default constructor (of `Foo<2>`)  for `Foo<1>`.

**Note**: Converting any of the `default` constructors (among `Foo<2>` and `S<*, 2>`) leads to expected assertion failure.
**Note**: `S<Foo<2>, 1>` works fine. Causes expected assertion failure.

**Comment by royjacobson** [comment]
I think it is likely that this is caused by the front-end thinking that `Foo<1>` is trivially default constructible and therefore eliminating the actual call to the constructor, hence the assert is ignored.

Hopefully it will be fixed when full support for P0848 is merged (https://reviews.llvm.org/D128619)

**Comment by usx95** [comment]
Awesome. I verified that your patch fixes the issue.

But I don't think this will eliminates the call to the constructor but actually chooses the wrong constructor. I came across this while working on https://reviews.llvm.org/D131479 and was able to see that we received the defaulted ctor of `Foo<2>` while instantiating `Foo<1>`.
Example:
```
template<unsigned N>
struct Foo {
  Foo() requires (N==1) {}
  Foo() requires (N==2) = default;
};

template <unsigned N, unsigned M>
struct S {
  Foo<M> data;
  S() requires (N==1) {}
  consteval S() requires (N==2) = default;
};

void func() {
  // Foo<1> default ctor is not consteval and cannot be used by S<2>.
  S<2, 1> must_fail;
  S<2, 2> should_work_fine;
}
```
This was a failing test for my patch (https://reviews.llvm.org/D131479). `S<2,1>` sees the `consteval` ctor for `Foo<2>` instead of `Foo<1>` and thus `S<2,1>` was incorrectly accepted.

**Comment by royjacobson** [comment]
The way the code works is that we usually go through Sema::LookupSpecialMember, which does overload resolution + constraints checks. It would be surprising if that had a bug. What I suspect is happening is that something checks HasTrivialConstructor and skips the actual constructor lookup when it (wrongly) returns true.

**Comment by AaronBallman** [comment]
Also possibly worth remembering: when Clang is building up a class declaration, we incrementally track the members that are added to set a bunch of flags on the class for things like "has trivial default constructor". We have overload resolution handlers for copy and move constructors, but we don't for default constructors: https://github.com/llvm/llvm-project/blob/main/clang/include/clang/AST/CXXRecordDeclDefinitionBits.def#L145

**Comment by usx95** [comment]
Fixed by 7171615099142ed49042c0f27af437b05150dd9b
